File: 749-shortest-completing-word/shortest-completing-word.swift
class Solution {
    func shortestCompletingWord(_ licensePlate: String, _ words: [String]) -> String {
        var words = words
        var wordsSortedByLength = words.sorted(by: {$0.count < $1.count})

        var extractedString = licensePlate.filter { character in !character.isWhitespace && !character.isNumber }
                                        .lowercased()

        var remainding = [String]()
    
        for var word in wordsSortedByLength {
            word.removeAll(where: { chac in
                !extractedString.contains(chac)
            })
            remainding.append(word)
        }
    
        let diff = zip(wordsSortedByLength, remainding)
    
        for (_,value) in diff.enumerated() {
            if extractedString.sorted() == value.1.sorted() || extractedString.sorted().count < value.1.sorted().count {
                return value.0
            }
        }
        return " "
    }
}

File: 749-shortest-completing-word/shortest-completing-word.kt
class Solution {
    fun shortestCompletingWord(licensePlate: String, words: Array<String>): String {
        val map: Map<Char, Int> = Parser(licensePlate).invoke()
        words.sortBy { it.length }
        for(word in words){
            if(isCompletable(word, map)){
                return word
            }
        }
        return ""
    }

    private fun isCompletable(word: String, map: Map<Char, Int>): Boolean {
        val wordMap = Parser(word).invoke()
        for((key, value) in map){
            if(wordMap.getOrDefault(key, 0) < value){
                return false
            }
        }
        return true
    }
}

data class Parser(val text :String){
    fun invoke(): Map<Char, Int> {
        val map = HashMap<Char, Int>()
        for (i in text){
            if(i.isLetter()){
                val char = i.lowercaseChar()
                map[char] = map.getOrDefault(char, 0) + 1
            }
        }
        return map
    }
}
